File: src/Program/Sort.hs
module Program.Sort where

import Prelude hiding (succ, min, max)
import Syntax
import Program.Bool
import Program.Num
import Program.List hiding (a, b)

minmaxo :: G a -> G a
minmaxo g =
  Let 
    ( def "minmaxo" ["a", "b", "min", "max"] 
      (
        (min === a &&& max === b &&& call "leo" [a, b, trueo]) |||
        (max === a &&& min === b &&& call "gto" [a, b, trueo])
      )
    ) $ leo $ gto g
    where [a, b, min, max] = map V ["a", "b", "min", "max"]

smallesto :: G a -> G a
smallesto g =
  Let 
    ( def "smallesto" ["l", "s", "l'"] 
      (
        l === s % nil &&& l' === nil |||
        fresh ["h", "t", "s'", "t'", "max"] 
        (
          l' === max % t' &&&
          l === h % t &&&
          call "minmaxo" [h, s', s, max] &&&
          call "smallesto" [t, s', t']
        )
      )
    ) $ minmaxo g
    where [l, s, l', h, t, s', t', max] = map V ["l", "s", "l'", "h", "t", "s'", "t'", "max"]

sorto :: G a -> G a
sorto g =
  Let 
    ( def "sorto" ["x", "y"] 
      (
        x === nil &&& y === nil |||
        fresh ["s", "xs", "xs'"]
          ( 
            y === s % xs' &&& 
            call "sorto" [xs, xs'] &&& 
            call "smallesto" [x, s, xs] 
          )
      )
    ) $ smallesto g
    where [x, y, s, xs, xs'] = map V ["x", "y", "s", "xs", "xs'"]

File: papers/miniKanren/ocanren/sort/sort.pl
smallesto(cons(Y3, nil), Y3, nil).
smallesto(cons(Q3, Q4), Y3, cons(Q1, Q2)) :- minmaxo(Y3, Q3, Q5, Q1), smallesto(Q4, Q5, Q2).
sorto(nil, nil).
sorto(Y5, cons(Q1, Q2)) :- sorto(Q3, Q2), smallesto(Y5, Q1, Q3).
minmaxo(o, o, Y10, Y10).
minmaxo(s(Q1), s(Q1), s(Q2), s(Q2)) :- leo(Q1, Q2).
minmaxo(o, s(Q3), o, s(Q3)).
minmaxo(s(Q5), s(Q4), s(Q5), s(Q4)) :- gto(Q5, Q4).
leo(o, Y12).
leo(s(Q1), s(Q2)) :- leo(Q1, Q2).
gto(o, s(Q1)).
gto(s(Q3), s(Q2)) :- gto(Q3, Q2).
